The built-in adapter for Falcon framework does not work with Python 3.11

As of Nov 2022, the latest version of Falcon framework is not yet compatible with Python 3.11. When you try to run your code built with Falcon, Python 3.11 raises the following TypeError:

```
_______ ERROR collecting tests/adapter_tests_async/test_async_falcon.py ________
ImportError while importing test module '/home/runner/work/bolt-python/bolt-python/tests/adapter_tests_async/test_async_falcon.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/adapter_tests_async/test_async_falcon.py:11: in <module>
    from slack_bolt.adapter.falcon.async_resource import AsyncSlackAppResource
slack_bolt/adapter/falcon/async_resource.py:5: in <module>
    from falcon.asgi import Request, Response
/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/site-packages/falcon/asgi/__init__.py:35: in <module>
    from .app import App
/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/site-packages/falcon/asgi/app.py:46: in <module>
    from .response import Response
/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/site-packages/falcon/asgi/response.py:17: in <module>
    from asyncio.coroutines import CoroWrapper  # type: ignore
E   ImportError: cannot import name 'CoroWrapper' from 'asyncio.coroutines' (/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/asyncio/coroutines.py)
```

There is no workaround until Falcon project releases a new version with the fix on their side.

See also:
 * https://github.com/slackapi/bolt-python/pull/751#issuecomment-1293239137
 * https://github.com/falconry/falcon/issues/2120

### Category (place an `x` in each of the `[ ]`)

* [ ] **slack_bolt.App** and/or its core components
* [ ] **slack_bolt.async_app.AsyncApp** and/or its core components
* [x] Adapters in **slack_bolt.adapter**
* [ ] Others

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -48,15 +48,13 @@ jobs:
       - name: Run tests for HTTP Mode adapters (Django)
         run: |
           pytest tests/adapter_tests/django/
-      # TODO: Enable testing with Falcon 4.x + Python 3.11
-      # See also: https://github.com/slackapi/bolt-python/issues/757
       - name: Run tests for HTTP Mode adapters (Falcon 3.x)
         run: |
-          if [ ${{ matrix.python-version }} != "3.11" ]; then
             pytest tests/adapter_tests/falcon/
-          fi
       - name: Run tests for HTTP Mode adapters (Falcon 2.x)
         run: |
+          # Falcon 2.x does not support Python 3.11 or newer
+          # See also: https://github.com/slackapi/bolt-python/issues/757
           if [ ${{ matrix.python-version }} != "3.11" ]; then
             pip install "falcon<3"
             pytest tests/adapter_tests/falcon/
@@ -76,10 +74,6 @@ jobs:
       - name: Run tests for HTTP Mode adapters (asyncio-based libraries)
         run: |
           pip install -e ".[async]"
-          pip install "falcon>=3,<4"
-          if [ ${{ matrix.python-version }} == "3.11" ]; then
-            # TODO: Enable testing with Falcon 4.x + Python 3.11
-            # See also: https://github.com/slackapi/bolt-python/issues/757
-            rm -f tests/adapter_tests_async/test_async_falcon.py
-          fi
+          # Falcon supports Python 3.11 since its v3.1.1
+          pip install "falcon>=3.1.1,<4"
           pytest tests/adapter_tests_async/

--- a/setup.py
+++ b/setup.py
@@ -81,7 +81,7 @@ setuptools.setup(
             "click>=7,<8",  # for chalice
             "CherryPy>=18,<19",
             "Django>=3,<5",
-            "falcon>=2,<4",
+            "falcon>=3.11,<4" if sys.version_info.minor >= 11 else "falcon>=2,<4",
             "fastapi>=0.70.0,<1",
             "Flask>=1,<3",
             "Werkzeug>=2,<3",

--- a/setup.py
+++ b/setup.py
@@ -81,7 +81,7 @@ setuptools.setup(
             "click>=7,<8",  # for chalice
             "CherryPy>=18,<19",
             "Django>=3,<5",
-            "falcon>=3.11,<4" if sys.version_info.minor >= 11 else "falcon>=2,<4",
+            "falcon>=3.1.1,<4" if sys.version_info.minor >= 11 else "falcon>=2,<4",
             "fastapi>=0.70.0,<1",
             "Flask>=1,<3",
             "Werkzeug>=2,<3",

